src/uopool/database.rs

```diff
@@ -0,0 +1,328 @@
+    }
+
+    fn get(
+        &self,
+        user_operation_hash: &UserOperationHash,
+    ) -> Result<Option<UserOperation>, DBError> {
+        let tx = self.env.tx()?;
+        let res = tx.get::<UserOperationDB>(*user_operation_hash)?;
+        tx.commit()?;
+        Ok(res)
+    }
+
+    fn get_all_by_sender(&self, sender: &Address) -> Self::UserOperations {
+        let wrap_sender: WrapAddress = (*sender).into();
+        self.env
+            .tx()
+            .and_then(|tx| {
+                let mut cursor = tx.cursor_dup_read::<SenderUserOperationDB>()?;
+                let res = cursor
+                    .walk_dup(wrap_sender.clone(), Address::default().into())?
+                    .map(|a| a.map(|(_, v)| v))
+                    .collect::<Result<Vec<_>, _>>()?;
+                tx.commit()?;
+                Ok(res)
+            })
+            .unwrap_or_else(|_| vec![])
+    }
+
+    fn remove(&mut self, user_operation_hash: &UserOperationHash) -> Result<(), DBError> {
+        let tx = self.env.tx_mut()?;
+        if let Some(user_op) = tx.get::<UserOperationDB>(*user_operation_hash)? {
+            tx.delete::<UserOperationDB>(*user_operation_hash, None)?;
+            tx.delete::<SenderUserOperationDB>(user_op.sender.into(), Some(user_op))?;
+            tx.commit()?;
+            Ok(())
+        } else {
+            Err(DBError::NotFound)
+        }
+    }
+
+    #[cfg(debug_assertions)]
+    fn get_all(&self) -> Self::UserOperations {
+        self.env
+            .tx()
+            .and_then(|tx| {
+                let mut c = tx.cursor_read::<UserOperationDB>()?;
+                let res = c
+                    .walk(UserOperationHash::default())?
+                    .map(|a| a.map(|(_, v)| v))
+                    .collect::<Result<Vec<_>, _>>()?;
+                tx.commit()?;
+                Ok(res)
+            })
+            .unwrap_or_else(|_| vec![])
+    }
+
+    #[cfg(debug_assertions)]
+    fn clear(&mut self) {
+        self.env
+            .tx_mut()
+            .and_then(|tx| {
+                tx.clear::<UserOperationDB>()?;
+                tx.clear::<SenderUserOperationDB>()?;
+                tx.commit()
+            })
+            .expect("Clear database failed");
+    }
+}
+fn default_page_size() -> usize {
+    let os_page_size = page_size::get();
+
+    // source: https://gitflic.ru/project/erthink/libmdbx/blob?file=mdbx.h#line-num-821
+    let libmdbx_max_page_size = 0x10000;
+
+    // May lead to errors if it's reduced further because of the potential size of the
+    // data.
+    let min_page_size = 4096;
+
+    os_page_size.clamp(min_page_size, libmdbx_max_page_size)
+}
+
+impl<E: EnvironmentKind> UserOpDatabase<E> {
+    pub fn new(path: PathBuf) -> anyhow::Result<Self> {
+        let env = Environment::new()
+            .set_max_dbs(TABLES.len())
+            .set_geometry(Geometry {
+                size: Some(0..(1024 * 1024 * 1024 * 1024 * 4)), // TODO: reevaluate (4 tb)
```

Comment:
Are there any recommendations somewhere how to set these values?

Reply:
Some doc here https://erthink.github.io/libmdbx/group__c__settings.html#ga79065e4f3c5fb2ad37a52b59224d583e.

Based on my previous experiences on lmdb, the size here is upper bound of disk consumption and it depends on the available disk space of the machine. I saw many apps setting this value very large because the lib would throw errors if the size of the database reach to the value.